MudTable: Add SelectionChangeable parameter

Ben0421 opened the pull request:
<!--
MAKE SURE TO READ THE CONTRIBUTING GUIDE BEFORE CREATING A PR
https://github.com/MudBlazor/MudBlazor/blob/dev/CONTRIBUTING.md

Testing sections can be removed for documentation changes
-->

<!-- Provide a general summary of your changes in the Title above -->
<!-- Keep the title short and descriptive, as it will be used as a commit message -->

## Description
<!-- Describe your changes in detail and why. -->
<!-- Note any issues that are resolved by this PR -->
<!-- e.g. resolves #1337 or fixes #9310. -->
On certain cases, it is useful to temporary disable the checkboxes of a multi-selection table while keeping them in view. This can now be achieved by setting `MultiSelectionReadOnly` on a `MudTable`.

## How Has This Been Tested?
<!-- All PRs should implement unit tests if possible. -->
<!-- Please describe how you tested your changes. -->
<!-- Have you created new tests or updated existing ones? -->
<!-- e.g. unit | visually | none -->
- [x] Unit tests
- [x] Visually

## Type of Changes
<!-- What type of changes does your code introduce? Put an `x` in only one box that applies best: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation (fix or improvement to the website or code docs)

<!-- If you made any visual changes, provide screenshots of before/after. If it has moving parts, please attach a high quality video. -->
![image](https://github.com/MudBlazor/MudBlazor/assets/43224836/76f947a7-64ae-43be-a3c7-465160ef1c8a)

## Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] The PR is submitted to the correct branch (`dev`).
- [x] My code follows the code style of this project.
- [x] I've added relevant tests.

Commit:
MudTable: Add MultiSelectionReadOnly parameter

ScarletKuro commented on src/MudBlazor.Docs/Pages/Components/Table/Examples/TableMultiSelectExample.razor:
```diff
@@ -31,13 +29,15 @@
     private IEnumerable<Element> Elements = new List<Element>();
     private bool _selectOnRowClick = true;
-    private MudTable<Element> _table;
+	private bool _disableMultiSelection;
```
Incorrect indentation

ScarletKuro commented on src/MudBlazor.Docs/Pages/Components/Table/Examples/TableMultiSelectExample.razor:
```diff
@@ -31,13 +29,15 @@
     private bool _selectOnRowClick = true;
-    private MudTable<Element> _table;
+	private bool _disableMultiSelection;
+	private MudTable<Element> _table;
```
Incorrect indentation

ScarletKuro commented on src/MudBlazor.UnitTests.Viewer/TestComponents/Table/TableMultiSelectionTest9.razor:
```diff
@@ -0,0 +1,34 @@
+		new("911 RS", "Porsche", "GT3")
+	];
+	private TableGroupDefinition<TableGroupingTest.RacingCar> groups = new()
+    {
```
Incorrect indentation

ScarletKuro commented on src/MudBlazor/Components/Table/MudTFootRow.razor:
```diff
@@ -16,7 +16,7 @@
             @if (Checkable)
             {
-                <MudCheckBox @bind-Value="Checked" Class="mud-table-cell-checkbox"/>
+				<MudCheckBox @bind-Value="Checked" Disabled="CheckableReadOnly" Class="mud-table-cell-checkbox" />
```
Incorrect indentation

ScarletKuro commented on src/MudBlazor/Components/Table/MudTHeadRow.razor:
```diff
@@ -16,7 +16,7 @@
 			@if (Checkable)
 			{
-				<MudCheckBox @bind-Value="Checked" Class="mud-table-cell-checkbox" />
+				<MudCheckBox @bind-Value="Checked" Disabled="CheckableReadOnly" Class="mud-table-cell-checkbox" />
```
Incorrect indentation

ScarletKuro commented on src/MudBlazor/Components/Table/MudTable.razor:
```diff
@@ -45,7 +45,7 @@
                         else
                         {
-                            <MudTHeadRow Checkable="MultiSelection" Expandable="@(GroupBy?.Expandable ?? false)">
+								<MudTHeadRow Checkable="MultiSelection" CheckableReadOnly="MultiSelectionReadOnly" Expandable="@(GroupBy?.Expandable ?? false)">
```
Incorrect indentation

ScarletKuro commented on src/MudBlazor/Components/Table/MudTable.razor:
```diff
@@ -162,7 +163,7 @@
                         else
                         {
-                            <MudTFootRow Checkable="MultiSelection" Expandable="@(GroupBy?.Expandable ?? false)">
+								<MudTFootRow Checkable="MultiSelection" CheckableReadOnly="MultiSelectionReadOnly" Expandable="@(GroupBy?.Expandable ?? false)">
```
Incorrect indentation

ScarletKuro commented on src/MudBlazor/Components/Table/MudTr.razor:
```diff
@@ -40,7 +40,7 @@
                 @if (Checkable)
                 {
-                    <MudCheckBox @bind-Value="Checked" StopClickPropagation="true" Class="mud-table-cell-checkbox"></MudCheckBox>
+					<MudCheckBox @bind-Value="Checked" Disabled="CheckableReadOnly" StopClickPropagation="true" Class="mud-table-cell-checkbox"></MudCheckBox>
```
Incorrect indentation

ScarletKuro commented:
@henon do you have better name? To me `CheckableReadOnly` sounds weird.

henon commented:
Maybe `ReadOnlySelection`. To be honest this seems to be quite a fringe use-case. Can this not be achieved with CSS, i.e. by applying `pointer-events:none` to the checkboxes?

ScarletKuro commented:
> Maybe `ReadOnlySelection`. To be honest this seems to be quite a fringe use-case. Can this not be achieved with CSS, i.e. by applying `pointer-events:none` to the checkboxes?

@danielchalmers

Commit:
Fix indentation

henon commented:
Here you are: https://try.mudblazor.com/snippet/QOQoYVbbgLebzSho
![table disable selection](https://github.com/MudBlazor/MudBlazor/assets/44090/461bb252-6da7-434a-9c8f-dde44ad42f2f)

Ben0421 commented:
> Here you are: https://try.mudblazor.com/snippet/QOQoYVbbgLebzSho ![table disable selection](https://private-user-images.githubusercontent.com/44090/347950840-461bb252-6da7-434a-9c8f-dde44ad42f2f.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA3MjM3MjcsIm5iZiI6MTcyMDcyMzQyNywicGF0aCI6Ii80NDA5MC8zNDc5NTA4NDAtNDYxYmIyNTItNmRhNy00MzRhLTljOGYtZGRlNDRhZDQyZjJmLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzExVDE4NDM0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFiYTI0MGI4OTdlMDhlZDFhODIwNjEyYzJhZWNmNDE3OTRhNWI2YmJlYmI5NWFlNDJkNDI4NTEwNDYwNmJlYTQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.oNRKLkZ9lrMZ5eTbFqJpHp9fo88BltiFBjHMJt1l3Qc) [ ![table disable selection](https://private-user-images.githubusercontent.com/44090/347950840-461bb252-6da7-434a-9c8f-dde44ad42f2f.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA3MjM3MjcsIm5iZiI6MTcyMDcyMzQyNywicGF0aCI6Ii80NDA5MC8zNDc5NTA4NDAtNDYxYmIyNTItNmRhNy00MzRhLTljOGYtZGRlNDRhZDQyZjJmLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzExVDE4NDM0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFiYTI0MGI4OTdlMDhlZDFhODIwNjEyYzJhZWNmNDE3OTRhNWI2YmJlYmI5NWFlNDJkNDI4NTEwNDYwNmJlYTQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.oNRKLkZ9lrMZ5eTbFqJpHp9fo88BltiFBjHMJt1l3Qc) ](https://private-user-images.githubusercontent.com/44090/347950840-461bb252-6da7-434a-9c8f-dde44ad42f2f.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA3MjM3MjcsIm5iZiI6MTcyMDcyMzQyNywicGF0aCI6Ii80NDA5MC8zNDc5NTA4NDAtNDYxYmIyNTItNmRhNy00MzRhLTljOGYtZGRlNDRhZDQyZjJmLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzExVDE4NDM0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFiYTI0MGI4OTdlMDhlZDFhODIwNjEyYzJhZWNmNDE3OTRhNWI2YmJlYmI5NWFlNDJkNDI4NTEwNDYwNmJlYTQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.oNRKLkZ9lrMZ5eTbFqJpHp9fo88BltiFBjHMJt1l3Qc) [ ](https://private-user-images.githubusercontent.com/44090/347950840-461bb252-6da7-434a-9c8f-dde44ad42f2f.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA3MjM3MjcsIm5iZiI6MTcyMDcyMzQyNywicGF0aCI6Ii80NDA5MC8zNDc5NTA4NDAtNDYxYmIyNTItNmRhNy00MzRhLTljOGYtZGRlNDRhZDQyZjJmLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzExVDE4NDM0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFiYTI0MGI4OTdlMDhlZDFhODIwNjEyYzJhZWNmNDE3OTRhNWI2YmJlYmI5NWFlNDJkNDI4NTEwNDYwNmJlYTQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.oNRKLkZ9lrMZ5eTbFqJpHp9fo88BltiFBjHMJt1l3Qc)

You can still toggle a row's selection by using tab & space. A css solution is error prone and subject to browser compatibility issues; this is why I went this way.

henon commented:
You are right of course. We have `ReadOnly` on other components as well, such as `MudTreeView`. So I'd say let's go for a `ReadOnly` parameter instead of a specific "make only multi-selection read only" type specialized parameter.

Ben0421 commented:
> You are right of course. We have `ReadOnly` on other components as well, such as `MudTreeView`. So I'd say let's go for a `ReadOnly` parameter instead of a specific "make only multi-selection read only"-type specialized parameter.

Considering the row can be editable, the name `ReadOnly` could cause confusion as to it's intended usage (edit vs selection). But `ReadOnlySelection` does feels less weird.

The reasoning behind `CheckableReadOnly` was that it complements the `Checkable` parameter. If `ReadOnlySelection` is better, should `Checkable` be changed to `Selectable`?

henon commented:
The `Checkable` of `MudTr` is a good name because it determines whether it shows a check box. `MudTable` itself is inconsistent now that we changed `MudChipSet` and `MudTreeView` in v7 to use `SelectionMode="SelectionMode.MultiSelection"`. I will probably change `MudTable` to use `SelectionMode` too, in the near future. 

I see that we already have `ReadOnly` in `MudTable` which prevents inline editing but displays the inline editing field in a read only state. It would be quite awkward to add `ReadOnlyWhatever` to that.  

I would say we should name it `SelectionChangeable`, that will fit nicely with `SelectionMode`

henon commented:
PS: single selection should also be affected by `SelectionChangable="false"`.

Commit:
Change parameter name to SelectionChangeable and add row click test

Ben0421 commented:
> The `Checkable` of `MudTr` is a good name because it determines whether it shows a check box. `MudTable` itself is inconsistent now that we changed `MudChipSet` and `MudTreeView` in v7 to use `SelectionMode="SelectionMode.MultiSelection"`. I will probably change `MudTable` to use `SelectionMode` too, in the near future.

I am looking forward to the change in selection modes!

> I see that we already have `ReadOnly` in `MudTable` which prevents inline editing but displays the inline editing field in a read only state. It would be quite awkward to add `ReadOnlyWhatever` to that.
> 
> I would say we should name it `SelectionChangeable`, that will fit nicely with `SelectionMode`

I changed the names `MultiSelectionReadOnly` and `CheckableReadOnly` to `SelectionChangeable`.

henon reviewed (approved):
LGTM

henon commented:
@ScarletKuro this good to merge?

ScarletKuro commented on src/MudBlazor/Components/Table/MudTHeadRow.razor:
```diff
@@ -16,7 +16,7 @@
 			@if (Checkable)
 			{
-				<MudCheckBox @bind-Value="Checked" Class="mud-table-cell-checkbox" />
+                <MudCheckBox @bind-Value="Checked" ReadOnly="!SelectionChangeable" Disabled="!SelectionChangeable" Class="mud-table-cell-checkbox" />
```
This one has incorrect indention

henon committed:
indentation fix
